### DISCUSSION TITLE
"console: command not found"

### DISCUSSION THREAD

**Discussion by jamaya77** [ipxe/ipxe#293] [Q&A] [answered]
Hello,

I'm trying to use a background for my boot menu using these settings, but then generates a message "console: command not found" (see snapthot):

console --x 1024 --y 768 --picture http://192.168.140.21/menu/background.png --depth 32

cpair --foreground 7 --background 4 0

Can someone provide friendly instructions to resolve or a work around with this issue? Thank you for time.

![missingconsole](https://user-images.githubusercontent.com/80473349/111002705-e4423300-8353-11eb-9550-29a0c9fe972a.png)

**Comment by NiKiZe** [comment, marked as answer]
Documentation is at https://ipxe.org/cmd/console#build_options
The URL Http://Ipxe.org/2e022001 also mentions the build configuration, why did you not include the full error message, including the URL while you where typing it out?

**Reply by jamaya77** [reply]
Hi,
I attached a larger snapshot.

**Reply by mcb30** [reply]
@jamaya77 Your question has already been answered.  Please read the documentation at https://ipxe.org/cmd/console#build_options

**Comment by jamaya77** [comment]
I installed iPXE from repository (yum install ipxe-bootimgs), so when I look in the install directory (/usr/share/ipxe/), I don't see the folders or files I'm suppose to see per instructions.

However, I've now downloaded copy of iPXE (via git clone) and extracted it, so now I see the folders and files in question. Should I copy the missing files to the root of the install directory (/usr/share/ipxe/) and continue the instructions?

**Reply by mcb30** [reply]
I understand now. You will need to build iPXE from source in order to use the "console" command, since the feature is not enabled by default and so will not be present in the prebuilt binaries from the `ipxe-bootimgs` package.

The instructions at https://ipxe.org/download#source_code should get you started with building from source.

Which iPXE binary were you previously using from your `/usr/share/ipxe` directory?

**Comment by jamaya77** [comment]
I was using "ipxe-x86_64.efi" and  " undionly.kpxe". Later replaced "ipxe-x86_64.efi" with the "ipxe.efi" downloaded from http://boot.ipxe.org/ipxe.efi.

**Reply by mcb30** [reply]
OK.  You can rebuild those two using `make bin/undionly.kpxe bin-x86_64-efi/ipxe.efi`

**Comment by jamaya77** [comment]
I've installed of the required packages for iPXE with the exception of "liblzma" which wouldn't install for CentOS. However, to get the liblzma libraries, it was recommend to install "xz-devel" and "pcre-devel".

Edited the "general.h" file to enable console_cmd:
![generalh1](https://user-images.githubusercontent.com/80473349/111053679-c3510f00-8433-11eb-961f-97fcece3ec92.png)
![generalh2](https://user-images.githubusercontent.com/80473349/111053683-ca781d00-8433-11eb-8872-e4e779039d30.png)

Edited the "console.h":
![consoleh](https://user-images.githubusercontent.com/80473349/111053699-f4314400-8433-11eb-88ed-654755fa647f.png)

Ran the command without any errors, but no change "console command not found" :
make bin/undionly.kpxe bin-x86_64-efi/ipxe.efi

**Reply by mcb30** [reply]
OK, so if you have run this command:
```
make bin/undionly.kpxe bin-x86_64-efi/ipxe.efi
```
and the build completed with no errors, then you will now have the files `bin/undionly.kpxe` and `bin-x86_64-efi/ipxe.efi` within your source directory.  Where did you copy these files to in order to use them?

**Reply by jamaya77** [reply]
The /tftpboot directory is the source folder.

**Reply by mcb30** [reply]
You probably don't want to place the entire source tree in `/tftpboot`.  Build the code somewhere else (even on a completely different machine), then copy only the `undionly.kpxe` and `ipxe.efi` files to the location from which you were **already** serving the original versions of those files.

Before you started doing anything with the source code: where were your original copies of `undionly.kpxe` and `ipxe.efi` to be found?  (Most likely you will have previously put them in either `/tftpboot`, `/var/lib/tftpboot`, or `/srv/tftpboot` since those would be the most common locations from which to serve those files.)

**Comment by jamaya77** [comment]
Good thing I took a snapshot of the VM server before I built the code. The original copies of the files are located in /usr/share/ipxe directory, but the only original file I'm using is the "undionly.kpxe" file which was copied over to the "/tftpboot" directory. The "ipxe.efi" file was downloaded from http://boot.ipxe.org/ipxe.efi to the "/tftpboot" directory well.

I'll restore the environment back before I built the code. After the restore, if I clone the environment to rebuild the code, how should I proceed to include the rebuild of the "undionly.kpxe" and "ipxe.efi" files before I can copy those files?

**Reply by mcb30** [reply]
Thanks.  So: you can build the binaries on any machine.  There is no need to build the code on the specific VM that will be serving the files.

Once you have built your binaries using `make bin/undionly.kpxe bin-x86_64-efi/ipxe.efi`, all you need to do is to copy the files `bin/undionly.kpxe` and `bin-x86_64-efi/ipxe.efi` to the `/tftpboot` directory on your server.

**Comment by jamaya77** [comment]
Thank you!!, @mcb30 , I was able to rebuild the files (with the console command enabled) on another machine and copied them to the server. The backgound image for the boot menu loaded fine.

Thanks again!
